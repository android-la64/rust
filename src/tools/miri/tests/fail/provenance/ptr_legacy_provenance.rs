// compile-flags: -Zmiri-disable-stacked-borrows
// normalize-stderr-test: "offset -[0-9]+" -> "offset -XX"
#![feature(strict_provenance)]

use std::ptr;

// Make sure that with legacy provenance, the allocation id of
// a casted pointer is determined at cast-time
fn main() {
    let x: i32 = 0;
    let y: i32 = 1;

    let x_ptr = &x as *const i32;
    let y_ptr = &y as *const i32;

    let x_usize = x_ptr.expose_addr();
    let y_usize = y_ptr.expose_addr();

    let ptr = ptr::from_exposed_addr::<i32>(y_usize);
    let ptr = ptr.with_addr(x_usize);
    assert_eq!(unsafe { *ptr }, 0); //~ ERROR is out-of-bounds
}
