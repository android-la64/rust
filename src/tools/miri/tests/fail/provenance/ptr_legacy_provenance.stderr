error: Undefined Behavior: dereferencing pointer failed: ALLOC has size 4, so pointer to 4 bytes starting at offset -XX is out-of-bounds
  --> $DIR/ptr_legacy_provenance.rs:LL:CC
   |
LL |     assert_eq!(unsafe { *ptr }, 0);
   |                         ^^^^ dereferencing pointer failed: ALLOC has size 4, so pointer to 4 bytes starting at offset -XX is out-of-bounds
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `main` at $DIR/ptr_legacy_provenance.rs:LL:CC

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

